exporter/metrics.go

```diff
@@ -235,30 +235,45 @@ func metricHelp(prefix, name string) string {
 
 // buildMetrics is a wrapper around makeMetrics, because makeMetrics is recursive and requires a prefix
 // and a map of labels. From the collectors we call buildMetrics which has a simpler signature.
-func buildMetrics(m bson.M, compatibleMode bool) []prometheus.Metric {
-	return makeMetrics("", m, nil, compatibleMode)
+func buildMetrics(labels map[string]string, m bson.M, compatibleMode bool) []prometheus.Metric {
+	return makeMetrics("", m, labels, compatibleMode)
 }
 
 func makeMetrics(prefix string, m bson.M, labels map[string]string, compatibleMode bool) []prometheus.Metric {
 	var res []prometheus.Metric
 
+	originalLabels := func() map[string]string {
+		labelsCopy := map[string]string{}
+		for k, v := range labels {
+			labelsCopy[k] = v
+		}
+
+		return labelsCopy
+	}
+
+	// We need to use this copy of the original labels because while looping through the incoming map,
+	// we are adding more labels depending on each metric renaming rule but, we must clean the labels for
+	// the next metric being processed in the for looop
+	workingLabels := originalLabels()
+
 	if prefix != "" {
 		prefix += "."
 	}
 
 	for k, val := range m {
 		switch v := val.(type) {
 		case bson.M:
-			res = append(res, makeMetrics(prefix+k, v, labels, compatibleMode)...)
+			res = append(res, makeMetrics(prefix+k, v, workingLabels, compatibleMode)...)
 		case map[string]interface{}:
-			res = append(res, makeMetrics(prefix+k, v, labels, compatibleMode)...)
+			res = append(res, makeMetrics(prefix+k, v, workingLabels, compatibleMode)...)
 		case primitive.A:
 			v = []interface{}(v)
 			res = append(res, processSlice(prefix, k, v, compatibleMode)...)
 		case []interface{}:
 			continue
 		default:
-			rm, err := makeRawMetric(prefix, k, v, labels)
+			rm, err := makeRawMetric(prefix, k, v, workingLabels)
+			workingLabels = originalLabels() // Cleanup before processing a new metric
```

Comment:
Looks like we update labels only in makeRawMetric and processSlice. isn't it better to call `originalLabels` only before calling them?

Reply:
Removed this.